### PR TITLE
feat: Add Endpoint for Blog Category Creation

### PR DESCRIPTION
**Pull request by GreyyDaze**
**Description:**
This PR introduces a new endpoint for creating blog categories. It includes the implementation of the `BlogCategoryController`, `BlogCategoryService`, and related DTOs and entities.

**Testing Instructions:**
1. **Open Postman.**
2. **Set the request type to POST.**
3. **Enter the API URL:** `http://localhost:3000/api/v1/blog-categories`.
4. **Select the Body tab, choose raw, and set the format to JSON. Enter the following JSON for `CreateBlogCategoryDto`:**
**Example JSON:**
```json
{
  "name": "Tech"
} 
```
5. **Click Send.**
6. **Check the response:**
**Success:**
```json
{
   "status": "success",
   "message": "Blog category created successfully.",
   "data": {
       "name": "Tech"
    },
   "status_code": 201
}
```
**Errors:**
```json
{
    "status": "error",
    "message": "Category name already exists.",
    "status_code": 400
}
```
```json
{
  "status": "error",
  "message": "Failed to create category.",
  "status_code": 500
}
```
**Tasks Completed:**

- [X]  Added `BlogCategoryController` with a POST `/api/v1/blog-categories` endpoint.
- [X]  Created `CreateBlogCategoryDto` for input validation.
- [X]  Defined `BlogCategory` entity.
- [X]  Implemented error handling and response messages.
- [X]  Wrote unit and integration tests for the new functionality.

**Additional Notes:**

- Verify that category names are unique to avoid duplication issues.

**Related Issue:**

- Blog Category Creation #142

### CODE REVIEW

File: src/modules/blog/services/blog-category.service.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+      });
+    }
+
+    const category = this.blogCategoryRepository.create({ name });
+
+    try {
+      const savedCategory = await this.blogCategoryRepository.save(category);
+      return {
+        status: 'success',
+        message: 'Blog category created successfully.',
```
[the remark this answers is not in the document]
Comment (GreyyDaze) [reply src/modules/blog/services/blog-category.service.ts:33]: fix

File: src/modules/blog/services/blog-category.service.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import { BlogCategory } from '../entities/blog-category.entity';
+
+@Injectable()
+export class BlogCategoryService {
+  constructor(
+    @InjectRepository(BlogCategory)
+    private readonly blogCategoryRepository: Repository<BlogCategory>
+  ) {}
+
+  async createCategory(createBlogCategoryDto: CreateBlogCategoryDto): Promise<any> {
```
[the remark this answers is not in the document]
Comment (GreyyDaze) [reply src/modules/blog/services/blog-category.service.ts]: fix

File: src/modules/blog/dto/create-blog.dto.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
+import { IsNotEmpty, IsOptional, IsBoolean, IsString, IsNumber } from 'class-validator';
+
+export class CreateBlogDto {
+  @IsNotEmpty()
```
[the remark this answers is not in the document]
Comment (GreyyDaze) [reply src/modules/blog/dto/create-blog.dto.ts:5]: fix

File: src/modules/blog/controllers/blog.controller.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+import { BlogService } from '../services/blog.service';
+import { CreateBlogDto } from '../dto/create-blog.dto';
+import { Blog } from '../entities/blog.entity';
+
+@Controller('api/v1/blogs')
+export class BlogController {
+  constructor(private readonly blogService: BlogService) {}
+
+  @Post()
+  async create(
```
[the remark this answers is not in the document]
Comment (GreyyDaze) [reply src/modules/blog/controllers/blog.controller.ts]: fix

File: src/modules/blog/controllers/blog.controller.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    @Body() createBlogDto: CreateBlogDto
+  ): Promise<{ status: string; message: string; status_code: number; data?: Blog }> {
```
[the remark this answers is not in the document]
Comment (GreyyDaze) [reply src/modules/blog/controllers/blog.controller.ts]: fix

File: src/modules/blog/controllers/blog.controller.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
+import { Controller, Post, Body, HttpException, HttpStatus } from '@nestjs/common';
+import { BlogService } from '../services/blog.service';
+import { CreateBlogDto } from '../dto/create-blog.dto';
+import { Blog } from '../entities/blog.entity';
+
+@Controller('api/v1/blogs')
```
[the remark this answers is not in the document]
Comment (GreyyDaze) [reply src/modules/blog/controllers/blog.controller.ts]: fix

File: src/modules/blog/controllers/blog.controller.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (Cobidex) [reply src/modules/blog/controllers/blog.controller.ts]: Yes please

**Review by Cobidex — APPROVED** [review]
LGTM

**Review by Heba-WebDev — APPROVED** [review]

File: src/modules/blog/controllers/blog.controller.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (GreyyDaze) [reply src/modules/blog/controllers/blog.controller.ts]: fix

File: src/modules/blog/dto/blog-response.dto.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
+import { ApiProperty } from '@nestjs/swagger';
+import { Blog } from '../entities/blog.entity';
+import { UserType } from '../../../modules/user/entities/user.entity';
+import { Organisation } from '../../../modules/organisations/entities/organisations.entity';
+
+export class ResponseDto {
```
Comment (Homoakin619) [inline src/modules/blog/dto/blog-response.dto.ts:6]: rename to BlogResponseDto

File: src/modules/blog/controllers/blog.controller.ts [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+@Controller('/blogs')
+export class BlogController {
+  constructor(private readonly blogService: BlogService) {}
+
+  @Post()
+  @ApiOperation({ summary: 'Create a new blog' })
+  @ApiBody({ type: CreateBlogDto })
+  @ApiResponse({ status: 201, description: 'Blog created successfully', type: Blog })
+  @ApiResponse({ status: 500, description: 'Error creating blog' })
+  async create(@Body() createBlogDto: CreateBlogDto): Promise<ResponseDto> {
```
Comment (Homoakin619) [inline src/modules/blog/controllers/blog.controller.ts:18]: make this more descriptive like createBlogPost

File: src/modules/blog/entities/blog-category.entity.ts [hunk 9]
```diff
@@ -0,0 +1,12 @@
+import { Entity, Column, OneToMany } from 'typeorm';
+import { AbstractBaseEntity } from '../../../entities/base.entity';
+import { Blog } from './blog.entity';
+
+@Entity()
+export class BlogCategory extends AbstractBaseEntity {
```
Comment (Homoakin619) [inline src/modules/blog/entities/blog-category.entity.ts:6]: Rename this to BlogPostCategory and update others accordingly

File: src/modules/blog/entities/blog.entity.ts [hunk 10]
```diff
@@ -0,0 +1,29 @@
+import { Entity, Column, ManyToOne, OneToMany } from 'typeorm';
+import { AbstractBaseEntity } from '../../../entities/base.entity';
+import { User } from '../../user/entities/user.entity';
+import { BlogComment } from './blog-comments.entity';
+import { BlogCategory } from './blog-category.entity';
+
+@Entity()
+export class Blog extends AbstractBaseEntity {
```
Comment (Homoakin619) [inline src/modules/blog/entities/blog.entity.ts:8]: Rename to BlogPost

File: src/modules/blog/services/blog-category.service.ts [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+import { InjectRepository } from '@nestjs/typeorm';
+import { Repository } from 'typeorm';
+import { CreateBlogCategoryDto } from '../dto/create-blog-category.dto';
+import { BlogCategory } from '../entities/blog-category.entity';
+import { CategoryResponseDto } from '../dto/blog-category-response.dto';
+import { ApiTags, ApiOperation, ApiBody, ApiResponse } from '@nestjs/swagger';
+
+@Injectable()
+@ApiTags('categories')
+export class BlogCategoryService {
```
Comment (Homoakin619) [inline src/modules/blog/services/blog-category.service.ts:11]: rename to BlogPostCategoryService

File: src/modules/blog/services/blog-category.service.ts [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+    @InjectRepository(BlogCategory)
+    private readonly blogCategoryRepository: Repository<BlogCategory>
+  ) {}
+
+  @ApiOperation({ summary: 'Create a new blog category' })
+  @ApiBody({ type: CreateBlogCategoryDto })
+  @ApiResponse({ status: 201, description: 'Blog category created successfully.', type: CategoryResponseDto })
+  @ApiResponse({ status: 400, description: 'Category name already exists.' })
+  @ApiResponse({ status: 500, description: 'Failed to create category.' })
+  async createCategory(createBlogCategoryDto: CreateBlogCategoryDto): Promise<CategoryResponseDto> {
```
Comment (Homoakin619) [inline src/modules/blog/services/blog-category.service.ts:22]: rename to createBlogPostCategory

File: src/modules/blog/services/blog.service.ts [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+import { User } from '../../../modules/user/entities/user.entity';
+import { Blog } from '../entities/blog.entity';
+import { InjectRepository } from '@nestjs/typeorm';
+import { Repository } from 'typeorm';
+import { BlogMapper } from '../mappers/blog.mapper';
+import { ApiResponse, ApiTags, ApiOperation, ApiBody } from '@nestjs/swagger';
+
+@Injectable()
+@ApiTags('blogs')
+export class BlogService {
```
Comment (Homoakin619) [inline src/modules/blog/services/blog.service.ts:14]: rename to BlogPostService

File: src/modules/blog/services/blog.service.ts [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+
+    try {
+      const blog = this.blogRepository.create({ ...rest, author, category });
+      await this.blogRepository.save(blog);
+
+      const responseData = BlogMapper.toResponseDto(blog, author, category);
+
+      return {
+        status: 'success',
+        message: 'Blog created successfully',
```
Comment (Homoakin619) [inline src/modules/blog/services/blog.service.ts:50]: update to Blog post

File: src/modules/blog/dto/blog-response.dto.ts (hunk 7 above)
Comment (GreyyDaze) [reply src/modules/blog/dto/blog-response.dto.ts]: fix

File: src/modules/blog/controllers/blog.controller.ts (hunk 8 above)
Comment (GreyyDaze) [reply src/modules/blog/controllers/blog.controller.ts]: fix

File: src/modules/blog/entities/blog-category.entity.ts (hunk 9 above)
Comment (GreyyDaze) [reply src/modules/blog/entities/blog-category.entity.ts]: fix

File: src/modules/blog/entities/blog.entity.ts (hunk 10 above)
Comment (GreyyDaze) [reply src/modules/blog/entities/blog.entity.ts]: fix

File: src/modules/blog/services/blog-category.service.ts (hunk 11 above)
Comment (GreyyDaze) [reply src/modules/blog/services/blog-category.service.ts]: fix

File: src/modules/blog/services/blog-category.service.ts (hunk 12 above)
Comment (GreyyDaze) [reply src/modules/blog/services/blog-category.service.ts:23]: fix

File: src/modules/blog/services/blog.service.ts (hunk 14 above)
Comment (GreyyDaze) [reply src/modules/blog/services/blog.service.ts]: fix

File: src/modules/blog/services/blog.service.ts (hunk 13 above)
Comment (GreyyDaze) [reply src/modules/blog/services/blog.service.ts]: fix

**Review by Homoakin619 — DISMISSED** [review]

File: src/modules/blog/entities/blog-comments.entity.ts [hunk 15]
```diff
@@ -0,0 +1,17 @@
+import { Entity, PrimaryGeneratedColumn, Column, ManyToOne } from 'typeorm';
+import { BlogPost } from './blog.entity';
+import { AbstractBaseEntity } from '../../../entities/base.entity';
+import { User } from '../../user/entities/user.entity';
+
+@Entity()
+export class BlogPostComment extends AbstractBaseEntity {
+  @PrimaryGeneratedColumn()
```
Comment (Cobidex) [inline src/modules/blog/entities/blog-comments.entity.ts:8]: Remove this, since you're extending the base class and it's not the primary key column

Comment (GreyyDaze) [reply src/modules/blog/entities/blog-comments.entity.ts]: fix

**Pull request closed without merging**